arch/arm/core/thread.c

```diff
@@ -85,12 +85,16 @@ void z_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 #endif /* CONFIG_THREAD_USERSPACE_LOCAL_DATA */
 #endif /* CONFIG_USERSPACE */
 
-#if CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT && CONFIG_USERSPACE
-	/* This is required to work-around the case where the thread
-	 * is created without using K_THREAD_STACK_SIZEOF() macro in
-	 * k_thread_create(). If K_THREAD_STACK_SIZEOF() is used, the
-	 * Guard size has already been take out of stackSize.
+#if defined(CONFIG_USERSPACE) && \
+	defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
+	/* Guard is carved-out of the stack object, not reserved in advance.
+	 *
+	 * On non-power-of-two, no need to adjust pStackMem as it is already
+	 * in the right place from Z_THREAD_STACK_BUFFER(). No need to adjust
+	 * the size either, as extra size was added to the object for this
+	 * instead of being carved-out.
 	 */
+	pStackMem += MPU_GUARD_ALIGN_AND_SIZE;
```

Comment:
why are you increasing the pStackMem here, hasn't it been increased in line 58 already?

Reply:
It has not, because ARCH_THREAD_STACK_BUFFER() definition has changed. On a power-of-two system the guard is carved out, not reserved, so it is not added in the ARCH_THREAD_STACK_BUFFER() definition. We carve-it out here. I had hoped the preceding comment I added would explain this perhaps I need to revise it?